File: Perl/file_renamer.pl
#!/usr/bin/env perl

use strict;
use warnings;
use Getopt::Long qw(:config bundling);
use File::Spec;

my ($keep_ext, $dry_run, $start) = (1, 0, 1);
GetOptions(
    'no-keep-ext' => sub { $keep_ext = 0 },
    'extension=s' => \my $force_ext,
    'start=i'     => \$start,
    'n|dry-run'   => \$dry_run,
) or die_usage();

die_usage() unless @ARGV == 2;
my ($directory, $prefix) = @ARGV;
die "Not a directory: '$directory'\n" unless -d $directory;
die "--start must be >= 0\n" if $start < 0;

opendir(my $dh, $directory) or die "Could not open '$directory': $!\n";
my @files = sort grep { ! /^\.{1,2}$/ && -f File::Spec->catfile($directory, $_) }
            readdir($dh);
closedir($dh);

my $width   = length(scalar @files + $start - 1);
$width      = 3 if $width < 3;
my $counter = $start;

for my $file (@files) {
    my $ext = '';
    if (defined $force_ext) {
        $ext = $force_ext =~ /^\./ ? $force_ext : ".$force_ext";
    } elsif ($keep_ext && $file =~ /(\.[^.\/]+)$/) {
        $ext = $1;
    }
    my $new_name = sprintf('%s%0*d%s', $prefix, $width, $counter, $ext);
    my $src = File::Spec->catfile($directory, $file);
    my $dst = File::Spec->catfile($directory, $new_name);

    if ($src eq $dst) { $counter++; next }
    if (-e $dst) {
        warn "Skipping '$file': '$new_name' already exists\n";
        $counter++;
        next;
    }

    if ($dry_run) {
        print "DRY-RUN: $file -> $new_name\n";
    } else {
        rename $src, $dst or die "Could not rename '$file' -> '$new_name': $!\n";
    }
    $counter++;
}

print "Files renamed successfully.\n" unless $dry_run;

sub die_usage {
    die "Usage: $0 [--no-keep-ext] [--extension EXT] [--start N] [-n] <directory> <prefix>\n";
}

__END__

# Example Commands:
# 1. Rename all files in the `photos` directory with the prefix `image_` (preserving extensions):
#    perl file_renamer.pl photos image_
#
# 2. Force a `.txt` extension on all renamed files:
#    perl file_renamer.pl --extension txt documents doc_
#
# 3. Preview the rename without applying it:
#    perl file_renamer.pl -n reports report_

File: Python3/file_renamer.py
#!/usr/bin/env python3

"""Prefix every regular file in a directory with a chosen string."""

import argparse
import sys
from pathlib import Path


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("directory", type=Path, help="Directory containing files to rename.")
    parser.add_argument("prefix", help="Prefix to prepend (a single underscore is added between prefix and original name).")
    parser.add_argument(
        "-n",
        "--dry-run",
        action="store_true",
        help="Show what would happen without renaming anything.",
    )
    return parser.parse_args()


def main() -> int:
    args = parse_args()
    if not args.directory.is_dir():
        print(f"Error: not a directory: {args.directory}", file=sys.stderr)
        return 1

    prefix = args.prefix
    files = sorted(p for p in args.directory.iterdir() if p.is_file())
    if not files:
        print("No regular files to rename.")
        return 0

    renamed = 0
    for src in files:
        if src.name.startswith(f"{prefix}_"):
            continue  # Already renamed; idempotent re-run.
        dst = src.with_name(f"{prefix}_{src.name}")
        if dst.exists():
            print(f"Skipping '{src.name}': '{dst.name}' already exists.", file=sys.stderr)
            continue
        if args.dry_run:
            print(f"DRY-RUN: {src.name} -> {dst.name}")
        else:
            src.rename(dst)
            renamed += 1
    if not args.dry_run:
        print(f"Renamed {renamed} file(s) in {args.directory}.")
    return 0


if __name__ == "__main__":
    sys.exit(main())
